Split plans

pommi opened the pull request:
Because of some Amazon limitations ([100 buckets](http://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html), [5000 IAM Users](http://docs.aws.amazon.com/IAM/latest/UserGuide/limits.html)), we at Mendix have been working on a "[Shared S3 Plan](https://github.com/mendix/s3-cf-service-broker/tree/shared-plan)", next to the Basic S3 Plan, where we share 1 bucket and 1 shared IAM account with all "Service Instances".

In preparation of the additional "Shared S3 Plan", in this Pull Request we've done some refactoring to add support for a new plan more easily, mainly:
- Iam.java has been split up in iam/Iam.java and iam/plan/BasicPlanIam.java
- service/S3ServiceInstanceService.java is now more generic, Basic Plan specific stuff moved to plan/BasicPlan.java

ddo-mendix committed:
README.md: remove whitespace

pommi committed:
README.md: link to https://github.com/cloudfoundry-community/ instead of https://github.com/davidehringer/

pommi committed:
create a Plan interface and BasicPlan implementation to support multiple plans

The BasicPlan class contains basic plan specific code related to creating and
deleting ServiceInstances and ServiceInstanceBindings.

Commit:
pom.xml: force joda-time to version 2.8.1

Which is required when running on JDK 1.8u60 and newer.

Tests are failing with:
> com.amazonaws.services.s3.model.AmazonS3Exception: AWS authentication
> requires a valid Date or x-amz-date header

Reference: https://github.com/aws/aws-sdk-java/commit/30b3d16

xiwenc commented:
@davidehringer Could you review this PR and merge if it's acceptable? It looks good to me.

davidehringer commented on src/main/java/org/cloudfoundry/community/servicebroker/s3/service/s3/S3.java:
```diff
@@ -25,30 +26,21 @@
     private static final Logger logger = LoggerFactory.getLogger(S3.class);
 
-    private final AmazonS3 s3;
+    private final AmazonS3 amazonS3;
```
It looks like the only change in this file is renaming this instance variable. Is this necessary as part of this as it adds quite a bit of noise to the commit?

davidehringer commented:
@pommi I'm not totally sure where you are headed with future commits but it seems like with the current changes there is going to be a lot of parallel hierarchies and duplicated logic.  I'll try to ask some questions in the commits.

davidehringer commented on src/main/java/org/cloudfoundry/community/servicebroker/s3/service/S3ServiceInstanceBindingService.java:
```diff
@@ -15,65 +15,44 @@
-
-    private final S3 s3;
-    private final Iam iam;
+    private final BasicPlan basicPlan;
```
If this is delegating to a `BasicPlan` it seems like you are going to have to exactly copy the `S3ServiceInstanceBindingService` to handle the shared plan concept.  Shouldn't this be some type of "Plan" interface that this service deals with?

pommi replied:
Thanks! Instead we've created a Plan interface in commit https://github.com/mendix/s3-cf-service-broker/commit/cf4a4e7

davidehringer commented on src/main/java/org/cloudfoundry/community/servicebroker/s3/service/S3ServiceInstanceService.java:
```diff
@@ -15,62 +15,50 @@
-
-    private final Iam iam;
-    private final S3 s3;
+    private final BasicPlan basicPlan;
```
Same comment as above regarding this using BasicPlan.

pommi replied:
See my comment above.

davidehringer commented on src/main/java/org/cloudfoundry/community/servicebroker/s3/service/iam/Iam.java:
```diff
@@ -0,0 +1,142 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+
+public abstract class Iam {
```
Can you explain this approach?  It seems like `Iam` needs to be an interface containing some of the methods in this abstract class along with all the methods in `BasicPlanIam`.  No other classes should be reference `BasicPlanIam` but rather only the `Iam` interface. Otherwise it seems like `BasicPlan` and `BasicPlanIam` are very tightly coupled and shouldn't be in separate packages because they would never be used independently (basically the Common Closure and Common Reuse principles).

pommi replied:
To make this approach more clear, we've split up https://github.com/mendix/s3-cf-service-broker/commit/9378a54 from the old pull request into:
- https://github.com/mendix/s3-cf-service-broker/commit/cf4a4e7 [create a Plan interface and BasicPlan implementation to support multiple plans]
- https://github.com/mendix/s3-cf-service-broker/commit/f175b9c [split BasicPlanIam from Iam to support multiple plans]

In the 2nd commit we made `Iam.java` a more generic class and introduced `BasicPlanIam.java`. `getUserNameForBinding()`, `getGroupNameForInstance()` and `getPolicyNameForInstance()` moved there. The arguments of the functions in `Iam.java` don't contain `bindingId` or `instanceId` anymore. This logic also moved to `BasicPlanIam.java`.

davidehringer commented on src/main/java/org/cloudfoundry/community/servicebroker/s3/plan/BasicPlan.java:
```diff
@@ -0,0 +1,98 @@
+import java.util.Map;
+
+@Component
+public class BasicPlan {
```
There should probably be a `Plan` interface as all types of plans will need to implemented these methods and the Services in the `service` package shouldn't know about the different implementations.

pommi replied:
See above comment. We've created a Plan interface in commit https://github.com/mendix/s3-cf-service-broker/commit/cf4a4e7

pommi committed:
split BasicPlanIam from Iam to support multiple plans

In Iam.java use more generic functions. Don't pass instanceIds, which require
getXxxForInstance() or getXxxForBinding(). In BasicPlanIam.java use more
descriptive function names like createGroup**ForInstance** and
deleteUserAccessKeys**ForBinding**.

ddo-mendix committed:
restore LICENSE above all files

pommi committed:
README.md: restructure the order and improve readability

In "Creation and Naming of AWS Resources" first explain you have to create an
AWS user. After that describe the "Basic Plan" and as sub-items, describe the
"bucket policy" and "bucket tagging". Put testing at the bottom.

pommi commented:
@davidehringer The pull request has been updated, taking all your comments into account. I will reply to all the comments individually.

davidehringer commented:
Thanks @pommi !
